### PR TITLE
Add Intel CI

### PR DESCRIPTION
**Pull request by mathomp4**
This PR adds Intel oneAPI to the CI for curl. At the moment it doesn't work due to various Intel C issues with curl, but this PR might help figure those out?

### CODE REVIEW

**Comment by mathomp4** [comment]
> Your edits made the lines too long:
> 
> ```
> ./curl_sasl.c:359:88: warning: Longer than 79 columns (LONGLINE)
>     * non-pointer conversion from "int" to "unsigned short" may lose significant bits */
> ./doh.c:613:89: warning: Longer than 79 columns (LONGLINE)
>     *   non-pointer conversion from "int" to "unsigned char" may lose significant bits */
> ```

Oops. I can fix that if you like if you can tell me what's a good comment you'd like. (I guess I'm an "unlimited line length" sort of programmer. 😄 )

Or maybe there's a way to get you write access to my branch on my fork? Either is fine!

**Comment by bagder** [comment]
I also took the liberty to force-push after I rebased this branch on top of the recent ftp fix for icc.

**Comment by bagder** [comment]
Gosh, it really does not like getinfo.c...
~~~
getinfo.c(442): error #2259: non-pointer conversion from "unsigned long" to "double" may lose significant bits
        *param_doublep = (double)val;
                         ^
getinfo.c(451): error #2259: non-pointer conversion from "timediff_t={curl_off_t={long}}" to "double" may lose significant bits
      *param_doublep = DOUBLE_SECS(data->progress.timespent);
                       ^
getinfo.c(454): error #2259: non-pointer conversion from "timediff_t={curl_off_t={long}}" to "double" may lose significant bits
      *param_doublep = DOUBLE_SECS(data->progress.t_nslookup);
                       ^
make[2]: Leaving directory '/home/runner/work/curl/curl/lib'
getinfo.c(457): error #2259: non-pointer conversion from "timediff_t={curl_off_t={long}}" to "double" may lose significant bits
      *param_doublep = DOUBLE_SECS(data->progress.t_connect);
make[1]: Leaving directory '/home/runner/work/curl/curl/lib'
                       ^
getinfo.c(460): error #2259: non-pointer conversion from "timediff_t={curl_off_t={long}}" to "double" may lose significant bits
      *param_doublep = DOUBLE_SECS(data->progress.t_appconnect);
                       ^
...

File: lib/c-hyper.c [hunk 1]
```diff
@@ -1013,8 +1013,8 @@ CURLcode Curl_http(struct Curl_easy *data, bool *done)
     /* For HTTP/2, we show the Host: header as if we sent it, to make it look
        like for HTTP/1 but it isn't actually sent since :authority is then
        used. */
-    result = Curl_debug(data, CURLINFO_HEADER_OUT, data->state.aptr.host,
-                        strlen(data->state.aptr.host));
+    Curl_debug(data, CURLINFO_HEADER_OUT, data->state.aptr.host,
+               strlen(data->state.aptr.host));
     if(result)
```
Comment (jpmenil) [inline lib/c-hyper.c:1018]: i guess this check is not needed anynore?

Comment (bagder) [reply lib/c-hyper.c:1018]: Yes, nice catch. Thanks!

**Comment by mathomp4** [comment]
> The intel compiler warning 2259 (`non-pointer conversion from X to Y may lose significant bits`) is simply too much and goes beyond normal C programming, so I decided we ignore it globally, so once it all builds fine I will remove the individual changes done to cater for 2259.

@bagder Note that the way you are disabling the warnings with [`-wd`](https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-diagnostic-options/wd-qwd.html) and turning warnings into errors with [`-we`](https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-diagnostic-options/we-qwe.html) seems to be deprecated:
```
icc: command line remark #10010: option '-we140,147,165,266' is deprecated and will be removed in a future release. See '-help deprecated'
icc: command line remark #10010: option '-wd279,981,1469,2259' is deprecated and will be removed in a future release. See '-help deprecated'
```
Looking at their docs, it seems like the translation (on Linux/macOS) is:

- `-wd` → `-diag-disable`
- `-we` → `-diag-error`

**Comment by bagder** [comment]
@mathomp4 thanks, I failed to find which updated options to use myself so your help is appreciated. I'll switch to using them too as part of the work in this PR.

**Comment by bagder** [comment]
ok, I did the wrong `closes` PR number, have fixed in my local repo.

**Comment by bagder** [comment]
Thanks @mathomp4 !

**Pull request closed without merging**